src/main/kotlin/org/radarbase/push/integration/garmin/auth/GarminAuthValidator.kt

```diff
@@ -19,39 +20,65 @@ class GarminAuthValidator(
         return if (token.isBlank()) {
-            throw HttpUnauthorizedException("invalid_token", "The token was not empty")
+            throw HttpUnauthorizedException("invalid_token", "The token was empty")
         } else {
+            var isAnyUnauthorised = false
             // Enrich the request by adding the User
             // the data format in Garmin's post is { <data-type> : [ {<data-1>}, {<data-2>} ] }
             val tree = request.getProperty("tree") as JsonNode
-            val userId = tree[tree.fieldNames().next()][0][USER_ID_KEY]
-                ?: throw HttpUnauthorizedException("invalid_user", "No user id provided")
-            val user = try {
-                userRepository.findByExternalId(userId.asText())
-            } catch (exc: NoSuchElementException) {
-                throw HttpUnauthorizedException(
-                    "no_user", "The user could not be found in the " +
-                            "user repository"
-                )
-            }
 
-            if (!user.isAuthorized) {
-                throw HttpUnauthorizedException(
-                    "invalid_user", "The user does not seem to be authorized"
-                )
-            }
+            val userTreeMap: Map<User, JsonNode> =
+                // group by user ID since request can contain data from multiple users
+                tree[tree.fieldNames().next()].groupBy { node ->
+                    node[USER_ID_KEY].asText()
+                }.filter { (k, v) ->
+                    val accessToken = v[0][USER_ACCESS_TOKEN_KEY].asText()
+                    val user = try {
+                        userRepository.findByExternalId(k)
+                    } catch (exc: NoSuchElementException) {
+                        isAnyUnauthorised = true
+                        userRepository.deregisterUser(k, accessToken)
+                        logger.warn(
+                            "no_user: {}", "The user $k could not be found in the " +
+                                "user repository."
+                        )
+                        return@filter false
+                    }
+                    if (!user.isAuthorized) {
+                        userRepository.deregisterUser(user.serviceUserId, accessToken)
+                        isAnyUnauthorised = true
+                        logger.warn(
+                            "invalid_user: {}", "The user $k does not seem to be authorized."
+                        )
+                        return@filter false
+                    }
+                    if (userRepository.getAccessToken(user) != accessToken) {
+                        isAnyUnauthorised = true
+                        userRepository.deregisterUser(user.serviceUserId, accessToken)
+                        logger.warn(
+                            "invalid_token: {}", "The token for user $k does not" +
+                                " match with the records on the system."
```

Comment:
fixed text should always go in the message part. The `k` variable can be added as a `{}` variable.